oak_runtime/src/node/http/mod.rs

```diff
@@ -518,21 +494,131 @@ impl HttpResponseIterator {
+                                err
+                            });
+                        let header_value =
+                            http::header::HeaderValue::from_bytes(val).map_err(|err| {
+                                warn!("Error when parsing header value: {}", err);
+                                err
+                            });
+                        (header_name, header_value)
+                    })
+                    .filter(|(name, value)| name.is_ok() && value.is_ok())
+                    // can safely call unwrap since the `Error`s are filtered out
+                    .map(|(name, value)| (name.unwrap(), value.unwrap()))
+            })
+            .collect()
+    }
+}
+
+impl From<http::header::HeaderMap<http::header::HeaderValue>> for OakHeaderMap {
+    fn from(hyper_headers: http::header::HeaderMap<http::header::HeaderValue>) -> Self {
+        #![allow(clippy::mutable_key_type)]
+        let mut headers = hashmap! {};
+        for (header_name, header_value) in &hyper_headers {
+            headers
+                .entry(header_name)
+                .or_insert_with(Vec::new)
+                .push(header_value.as_bytes().to_vec())
+        }
+        let headers = headers
+            .iter()
+            .map(|(name, value)| {
+                (
+                    name.to_string(),
+                    HeaderValue {
+                        values: value.to_owned(),
+                    },
+                )
+            })
+            .collect();
+
+        OakHeaderMap { headers }
+    }
+}
+
+/// Create an instance of Oak HttpRequest from the given hyper Request.
+async fn to_oak_http_request(req: Request<Body>) -> Result<HttpRequest, OakStatus> {
+    let uri = req.uri().to_string();
+    let method = req.method().as_str().to_string();
+    let headers = OakHeaderMap::from(req.headers().to_owned()).headers;
+    let body = hyper::body::to_bytes(req.into_body())
+        .await
+        .map_err(|err| {
+            warn!(
+                "Error when reading request body from the connection: {}",
+                err
+            );
+            OakStatus::ErrInternal
+        })?
+        .to_vec();
+
+    Ok(HttpRequest {
+        uri,
+        method,
+        body,
+        headers,
+    })
+}
+
+/// Convert an instance of Oak HttpResponse to hyper Response.
+fn to_hyper_http_response(http_response: HttpResponse) -> Result<Response<Body>, OakStatus> {
+    let mut builder = http::response::Builder::new();
+    let headers = OakHeaderMap {
+        headers: http_response.headers,
+    }
+    .to_vec();
+    for (header_name, header_value) in headers {
+        builder = builder.header(header_name, header_value);
+    }
+
+    builder
+        .status(http_response.status as u16)
+        .body(Body::from(http_response.body))
+        .map_err(|err| {
+            warn!("Could not build response: {}", err);
+            OakStatus::ErrInternal
+        })
+}
```

Comment:
Would https://doc.rust-lang.org/std/convert/trait.TryFrom.html or something similar help here?

Reply:
This has to be done in `oak_services`. I'll do this with #1459.